controllers/managedocs_controller.go

```diff
@@ -172,10 +172,7 @@ type ManagedOCSReconciler struct {
 // +kubebuilder:rbac:groups="odf.openshift.io",namespace=system,resources=storagesystems,verbs=list;watch;delete
 
 // SetupWithManager creates an setup a ManagedOCSReconciler to work with the provided manager
-func (r *ManagedOCSReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	ctrlOptions := controller.Options{
-		MaxConcurrentReconciles: 1,
-	}
+func (r *ManagedOCSReconciler) SetupWithManager(mgr ctrl.Manager, ctrlOptions controller.Options) error {
```

Comment:
Can we please have the controller option provided in the argument as an override options that should override the actual default options provided in this func?

Reply:
- sry, I partly understood it. You mean to have below as part of function but set the argument if supplied?
``` go
	ctrlOptions := controller.Options{
		MaxConcurrentReconciles: 1,
	}
```

Reply:
something like below
``` diff
diff --git a/controllers/managedocs_controller.go b/controllers/managedocs_controller.go
index 32a1203..44b4481 100644
--- a/controllers/managedocs_controller.go
+++ b/controllers/managedocs_controller.go
@@ -172,7 +172,12 @@ type ManagedOCSReconciler struct {
 // +kubebuilder:rbac:groups="odf.openshift.io",namespace=system,resources=storagesystems,verbs=list;watch;delete
 
 // SetupWithManager creates an setup a ManagedOCSReconciler to work with the provided manager
-func (r *ManagedOCSReconciler) SetupWithManager(mgr ctrl.Manager, ctrlOptions controller.Options) error {
+func (r *ManagedOCSReconciler) SetupWithManager(mgr ctrl.Manager, ctrlOptions *controller.Options) error {
+	if ctrlOptions == nil {
+		ctrlOptions = &controller.Options{
+			MaxConcurrentReconciles: 1,
+		}
+	}
 	managedOCSPredicates := builder.WithPredicates(
 		predicate.GenerationChangedPredicate{},
 	)
@@ -246,7 +251,7 @@ func (r *ManagedOCSReconciler) SetupWithManager(mgr ctrl.Manager, ctrlOptions co
 	)
 
 	return ctrl.NewControllerManagedBy(mgr).
-		WithOptions(ctrlOptions).
+		WithOptions(*ctrlOptions).
 		For(&v1.ManagedOCS{}, managedOCSPredicates).
 
 		// Watch owned resources
diff --git a/controllers/suite_test.go b/controllers/suite_test.go
index f4dca56..5123412 100644
--- a/controllers/suite_test.go
+++ b/controllers/suite_test.go
@@ -164,7 +164,7 @@ var _ = BeforeSuite(func() {
 			CustomerNotificationHTMLPath: testCustomerNotificationHTMLPath,
 		}
 
-		ctrlOptions := controller.Options{
+		ctrlOptions := &controller.Options{
 			MaxConcurrentReconciles: 1,
 			RateLimiter:             workqueue.NewItemFastSlowRateLimiter(5*time.Millisecond, 10*time.Second, 500),
 		}
diff --git a/main.go b/main.go
index 76c91f7..05d44fc 100644
--- a/main.go
+++ b/main.go
@@ -32,7 +32,6 @@ import (
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/client/config"
-	"sigs.k8s.io/controller-runtime/pkg/controller"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
 
 	"github.com/go-logr/logr"
@@ -117,10 +116,6 @@ func main() {
 		os.Exit(1)
 	}
 
-	ctrlOptions := controller.Options{
-		MaxConcurrentReconciles: 1,
-	}
-
 	addonName := envVars[addonNameEnvVarName]
 	if err = (&controllers.ManagedOCSReconciler{
 		Client:                       mgr.GetClient(),
@@ -137,7 +132,7 @@ func main() {
 		AlertSMTPFrom:                envVars[alertSMTPFromAddrEnvVarName],
 		DeploymentType:               envVars[deploymentTypeEnvVarName],
 		CustomerNotificationHTMLPath: "templates/customernotification.html",
-	}).SetupWithManager(mgr, ctrlOptions); err != nil {
+	}).SetupWithManager(mgr, nil); err != nil {
 		setupLog.Error(err, "Unable to create controller", "controller", "ManagedOCS")
 		os.Exit(1)
 	}
```

Reply:
Yes, that seems more reasonable, we will have default behavior and custom behavior options.

Reply:
tested and updated.